signer/src/blockchain_sync.rs

```diff
@@ -0,0 +1,336 @@
+) -> Result<(), Error> {
+    let term = ctx.get_termination_handle();
+    let stacks_client = ctx.get_stacks_client();
+    let storage = ctx.get_storage_mut();
+
+    let mut next_tenure_tip = *chain_tip;
+
+    loop {
+        if term.shutdown_signalled() {
+            return Ok(());
+        }
+
+        tracing::debug!(block_id = %next_tenure_tip, "syncing next tenure");
+
+        // Retrieve the tenure from the Stacks node representing the current
+        // Stacks chain-tip.
+        //
+        // Note: This specifically does not use the `fetch_unknown_ancestors`
+        // method as this is intended to be run on startup, which for an initial
+        // sync can result in high memory usage if all blocks from all tenures
+        // are fetched/buffered.
+        let tenure = stacks_client.get_tenure(next_tenure_tip).await?;
+
+        // If there are no blocks in the tenure (which shouldn't happen), then
+        // we fail the sync.
+        if tenure.blocks().is_empty() {
+            tracing::error!("received an empty tenure from the stacks node, failing sync");
+            return Err(Error::EmptyTenure(next_tenure_tip));
+        }
+
+        // Retrieve the anchor Bitcoin block from our local storage. We should
+        // have already processed this block and thus it should exist.
+        let bitcoin_block = storage.get_bitcoin_block(&tenure.anchor_block_hash).await?;
+
+        // If the Bitcoin block is missing, we fail the sync because then our
+        // stacks-block-linking is broken.
+        let Some(bitcoin_block) = bitcoin_block else {
+            tracing::error!("missing anchor block, failing sync");
+            return Err(Error::MissingBitcoinBlock(tenure.anchor_block_hash));
+        };
+
+        // We use `get_tenure` to fetch all of the Nakamoto blocks for each
+        // tenure, walking backwards, which accepts a `StacksBlockId` as its
+        // parameter. As we're walking backwards, we grab the first block in
+        // the tenure and use its parent block ID for the next iteration, which
+        // should give us the previous tenure and all of its blocks.
+        let first_block = tenure
+            .blocks()
+            .first()
+            .ok_or(Error::EmptyTenure(next_tenure_tip))?;
+
+        next_tenure_tip = first_block.header.parent_block_id;
+
+        // We re-use the same code as the block observer to write the blocks to
+        // the database.
+        block_observer::write_stacks_blocks(
+            &ctx.get_storage_mut(),
+            &ctx.config().signer.deployer,
+            &[tenure],
+        )
+        .await?;
+
+        // If the block height of the anchor block is equal to the nakamoto
+        // activation height then we're done.
+        if bitcoin_block.block_height == nakamoto_activation_height {
+            tracing::info!("nakamoto activation height reached, stacks block-sync completed");
+            return Ok(());
+        }
```

Comment:
```suggestion
        // If the block height of the anchor block is equal to the nakamoto
        // activation height then we're done.
        if bitcoin_block.block_height <= nakamoto_activation_height {
            tracing::info!("nakamoto activation height reached, stacks block-sync completed");
            return Ok(());
        }
```
since we can have gaps? also to be consistent with `fetch_unknown_ancestors`

Reply:
Yeah good point 👍